File: src/cursor.ts
export interface ChannelHistoryCursor {
  v: 1
  scope: 'channel'
  channelId: string
  boundary: { createAt: number; id: string }
  since: number | null
  safeBeforePostId?: string
}

const MAX_ENCODED_CURSOR_LENGTH = 2048
const MAX_DECODED_CURSOR_LENGTH = 1536
const MAX_DATE_MILLISECONDS = 8_640_000_000_000_000
const MAX_ID_LENGTH = 128
const SAFE_ID = /^[A-Za-z0-9_-]+$/

function invalidCursor(): never {
  throw new Error('Invalid cursor.')
}

function exactKeys(value: Record<string, unknown>, required: string[], optional: string[] = []) {
  const allowed = new Set([...required, ...optional])
  return (
    required.every((key) => Object.hasOwn(value, key)) &&
    Object.keys(value).every((key) => allowed.has(key))
  )
}

function isSafeId(value: unknown): value is string {
  return (
    typeof value === 'string' &&
    value.length > 0 &&
    value.length <= MAX_ID_LENGTH &&
    SAFE_ID.test(value)
  )
}

function validateCursor(value: unknown): ChannelHistoryCursor {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return invalidCursor()
  const candidate = value as Record<string, unknown>
  if (
    !exactKeys(candidate, ['v', 'scope', 'channelId', 'boundary', 'since'], ['safeBeforePostId'])
  ) {
    return invalidCursor()
  }
  const boundary = candidate.boundary
  if (!boundary || typeof boundary !== 'object' || Array.isArray(boundary)) return invalidCursor()
  const boundaryValue = boundary as Record<string, unknown>
  if (!exactKeys(boundaryValue, ['createAt', 'id'])) return invalidCursor()
  if (
    candidate.v !== 1 ||
    candidate.scope !== 'channel' ||
    !isSafeId(candidate.channelId) ||
    !isSafeId(boundaryValue.id) ||
    typeof boundaryValue.createAt !== 'number' ||
    !Number.isSafeInteger(boundaryValue.createAt) ||
    boundaryValue.createAt < 0 ||
    boundaryValue.createAt > MAX_DATE_MILLISECONDS ||
    !(
      candidate.since === null ||
      (typeof candidate.since === 'number' &&
        Number.isSafeInteger(candidate.since) &&
        candidate.since >= 0 &&
        candidate.since <= boundaryValue.createAt)
    ) ||
    (Object.hasOwn(candidate, 'safeBeforePostId') && !isSafeId(candidate.safeBeforePostId))
  ) {
    return invalidCursor()
  }
  return value as ChannelHistoryCursor
}

export function encodeChannelHistoryCursor(cursor: ChannelHistoryCursor): string {
  const validated = validateCursor(cursor)
  const encoded = Buffer.from(JSON.stringify(validated), 'utf8').toString('base64url')
  if (encoded.length > MAX_ENCODED_CURSOR_LENGTH) return invalidCursor()
  return encoded
}

export function decodeChannelHistoryCursor(encoded: string): ChannelHistoryCursor {
  if (
    encoded.length === 0 ||
    encoded.length > MAX_ENCODED_CURSOR_LENGTH ||
    !/^[A-Za-z0-9_-]+$/.test(encoded)
  ) {
    return invalidCursor()
  }
  let text: string
  try {
    const bytes = Buffer.from(encoded, 'base64url')
    if (bytes.length === 0 || bytes.length > MAX_DECODED_CURSOR_LENGTH) return invalidCursor()
    text = bytes.toString('utf8')
    if (Buffer.from(text, 'utf8').toString('base64url') !== encoded) return invalidCursor()
  } catch {
    return invalidCursor()
  }
  let value: unknown
  try {
    value = JSON.parse(text)
  } catch {
    return invalidCursor()
  }
  return validateCursor(value)
}

export function comparePostIds(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0
}

File: internal/cursor/cursor.go
// Package cursor encodes and decodes opaque channel-history cursors.
package cursor

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"io"
	"math"
	"regexp"
	"unicode/utf8"
)

const (
	maxEncodedLength = 2048
	maxDecodedLength = 1536
	maxDateMillis    = int64(8_640_000_000_000_000)
	maxIDLength      = 128
)

var (
	// ErrInvalidCursor is returned for every invalid cursor without exposing its input.
	ErrInvalidCursor = errors.New("invalid cursor")
	safeIDPattern    = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)
)

// ChannelHistory identifies a deterministic boundary in one channel's history.
type ChannelHistory struct {
	Version          int      `json:"v"`
	Scope            string   `json:"scope"`
	ChannelID        string   `json:"channelId"`
	Boundary         Boundary `json:"boundary"`
	Since            *int64   `json:"since"`
	SafeBeforePostID string   `json:"safeBeforePostId,omitempty"`
}

// Boundary is the newest post included in a channel-history page.
type Boundary struct {
	CreateAt int64  `json:"createAt"`
	ID       string `json:"id"`
}

// EncodeChannelHistory returns canonical, unpadded base64url JSON.
func EncodeChannelHistory(value ChannelHistory) (string, error) {
	if !valid(value) {
		return "", ErrInvalidCursor
	}

	data, err := json.Marshal(value)
	if err != nil {
		return "", ErrInvalidCursor
	}
	encoded := base64.RawURLEncoding.EncodeToString(data)
	if len(encoded) > maxEncodedLength {
		return "", ErrInvalidCursor
	}
	return encoded, nil
}

// DecodeChannelHistory validates and decodes an opaque channel-history cursor.
func DecodeChannelHistory(encoded string) (ChannelHistory, error) {
	if len(encoded) == 0 || len(encoded) > maxEncodedLength || !safeIDPattern.MatchString(encoded) {
		return ChannelHistory{}, ErrInvalidCursor
	}

	data, err := base64.RawURLEncoding.DecodeString(encoded)
	if err != nil || len(data) == 0 || len(data) > maxDecodedLength || !utf8.Valid(data) ||
		base64.RawURLEncoding.EncodeToString(data) != encoded {
		return ChannelHistory{}, ErrInvalidCursor
	}

	value, ok := decodeJSON(data)
	if !ok || !valid(value) {
		return ChannelHistory{}, ErrInvalidCursor
	}
	return value, nil
}

// ComparePostIDs applies the bytewise ASCII ordering used for equal timestamps.
func ComparePostIDs(a, b string) int {
	if a < b {
		return -1
	}
	if a > b {
		return 1
	}
	return 0
}

func valid(value ChannelHistory) bool {
	return value.Version == 1 && value.Scope == "channel" &&
		isSafeID(value.ChannelID) && isSafeID(value.Boundary.ID) &&
		value.Boundary.CreateAt > 0 && value.Boundary.CreateAt <= maxDateMillis &&
		(value.Since == nil || (*value.Since >= 0 && *value.Since <= value.Boundary.CreateAt)) &&
		(value.SafeBeforePostID == "" || isSafeID(value.SafeBeforePostID))
}

func isSafeID(value string) bool {
	return len(value) >= 1 && len(value) <= maxIDLength && safeIDPattern.MatchString(value)
}

func decodeJSON(data []byte) (ChannelHistory, bool) {
	var outer map[string]json.RawMessage
	decoder := json.NewDecoder(bytes.NewReader(data))
	if err := decoder.Decode(&outer); err != nil || !onlyKeys(outer,
		[]string{"v", "scope", "channelId", "boundary", "since"}, []string{"safeBeforePostId"}) {
		return ChannelHistory{}, false
	}
	if err := decoder.Decode(&struct{}{}); !errors.Is(err, io.EOF) {
		return ChannelHistory{}, false
	}

	var boundary map[string]json.RawMessage
	if err := json.Unmarshal(outer["boundary"], &boundary); err != nil ||
		!onlyKeys(boundary, []string{"createAt", "id"}, nil) {
		return ChannelHistory{}, false
	}

	version, ok := integer(outer["v"])
	if !ok || version != 1 {
		return ChannelHistory{}, false
	}
	createAt, ok := integer(boundary["createAt"])
	if !ok {
		return ChannelHistory{}, false
	}

	value := ChannelHistory{Version: int(version), Boundary: Boundary{CreateAt: createAt}}
	if json.Unmarshal(outer["scope"], &value.Scope) != nil ||
		json.Unmarshal(outer["channelId"], &value.ChannelID) != nil ||
		json.Unmarshal(boundary["id"], &value.Boundary.ID) != nil {
		return ChannelHistory{}, false
	}
	if !bytes.Equal(outer["since"], []byte("null")) {
		since, valid := integer(outer["since"])
		if !valid {
			return ChannelHistory{}, false
		}
		value.Since = &since
	}
	if raw, exists := outer["safeBeforePostId"]; exists {
		if json.Unmarshal(raw, &value.SafeBeforePostID) != nil || value.SafeBeforePostID == "" {
			return ChannelHistory{}, false
		}
	}
	return value, true
}

func integer(raw json.RawMessage) (int64, bool) {
	var number float64
	if json.Unmarshal(raw, &number) != nil || math.IsNaN(number) || math.IsInf(number, 0) ||
		math.Trunc(number) != number || number < -9_007_199_254_740_991 || number > 9_007_199_254_740_991 {
		return 0, false
	}
	return int64(number), true
}

func onlyKeys(value map[string]json.RawMessage, required, optional []string) bool {
	if value == nil || len(value) < len(required) || len(value) > len(required)+len(optional) {
		return false
	}
	allowed := make(map[string]struct{}, len(required)+len(optional))
	for _, key := range append(required, optional...) {
		allowed[key] = struct{}{}
	}
	for _, key := range required {
		if _, exists := value[key]; !exists {
			return false
		}
	}
	for key := range value {
		if _, exists := allowed[key]; !exists {
			return false
		}
	}
	return true
}
